### PR TITLE
impl(storage): add a OTel tracing client

### PR DESCRIPTION
**Pull request by coryan**
This will inject useful spans in most cases. I expect some of these spans will be replaced by spans for pagination wrappers, for `ReadObject()`, and for `WriteObject()`. But most of them are useful, and they provide more visibility into the library behavior.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11390)

### CODE REVIEW

File: google/cloud/storage/internal/tracing_client.cc [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+  auto scope = opentelemetry::trace::Scope(span);
+  return internal::EndSpan(*span, impl_->GetObjectMetadata(request));
+}
+
+StatusOr<std::unique_ptr<storage::internal::ObjectReadSource>>
+TracingClient::ReadObject(
+    storage::internal::ReadObjectRangeRequest const& request) {
+  auto span = internal::MakeSpan("storage::Client::ReadObject");
+  auto scope = opentelemetry::trace::Scope(span);
+  // TODO(#.....) - add a wrapper for ReadObjectSource.
```
Comment (dbolduc) [inline google/cloud/storage/internal/tracing_client.cc:130]: I am surprised `clang-tidy` let you get away with this.

**Review by dbolduc — APPROVED** [review]
Cool!

Were the `RawClient` span names intentional?

File: google/cloud/storage/internal/tracing_client.cc [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+    storage::internal::RewriteObjectRequest const& request) {
+  auto span = internal::MakeSpan("storage::Client::RewriteObject");
+  auto scope = opentelemetry::trace::Scope(span);
+  return internal::EndSpan(*span, impl_->RewriteObject(request));
+}
+
+StatusOr<storage::internal::CreateResumableUploadResponse>
+TracingClient::CreateResumableUpload(
+    storage::internal::ResumableUploadRequest const& request) {
+  auto span = internal::MakeSpan("storage::RawClient::CreateResumableUpload");
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/tracing_client.cc]: There are no `storage::Client::CreateResumableUpload` function, but I think the name I picked was bad. PTAL.

File: google/cloud/storage/internal/tracing_client.cc [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+  auto scope = opentelemetry::trace::Scope(span);
+  return internal::EndSpan(*span, impl_->CreateResumableUpload(request));
+}
+
+StatusOr<storage::internal::QueryResumableUploadResponse>
+TracingClient::QueryResumableUpload(
+    storage::internal::QueryResumableUploadRequest const& request) {
+  auto span = internal::MakeSpan("storage::RawClient::QueryResumableUpload");
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/tracing_client.cc]: intentional, but not a good idea.

File: google/cloud/storage/internal/tracing_client.cc [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+    storage::internal::DeleteResumableUploadRequest const& request) {
+  auto span = internal::MakeSpan("storage::Client::DeleteResumableUpload");
+  auto scope = opentelemetry::trace::Scope(span);
+  return internal::EndSpan(*span, impl_->DeleteResumableUpload(request));
+}
+
+StatusOr<storage::internal::QueryResumableUploadResponse>
+TracingClient::UploadChunk(
+    storage::internal::UploadChunkRequest const& request) {
+  auto span = internal::MakeSpan("storage::RawClient::UploadChunk");
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/tracing_client.cc]: Ditto.

File: google/cloud/storage/internal/tracing_client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/tracing_client.cc]: Fixed. `clang-tidy` and `checkers` do not know anything about TODO entries.

File: google/cloud/storage/internal/tracing_client.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/storage/internal/tracing_client.cc]: Good idea to have a consistent prefix.

**Pull request merged**